### ISSUE TITLE
Cannot spawn new connections.

### ISSUE THREAD

**Issue by tuboythers** [hyperium/h2#441]
Hi, I may be missing something, I've read through the documentation, as well as some of the issues, but I stumbled upon a matter which I cannot get past through.

So, I have the following piece of code: 
```
let mut h2 = server::handshake(tls_stream).await;

match h2 {
        Ok(mut h2) => {
                 while let Some(request) = h2.accept().await {
                          let (request, mut respond) = request.unrwap();

                          let response = Response::builder().status(StatusCode::OK).body(()).expect("Err building response");

                          let mut send = respond.send_response(response, false).expect("Err send_response");

                          let contents = b"index.html contents".to_vect();
                          send.send_data(contents.into(), true).expect("Err send_data");
                 }
        },
        Err(e) => error!("H2 Error: {}", e)
}
```

And this generally works, I can fire up a browser tab, open the local host address, connect via port 443/https and it will print the string `index.html contents` in the browser. However, this results to another issue, I cannot open another tab. It just sits there loading forever until I kill the tab or the server. The same happens over cURL too, and nothing else happens. If I close the original tab, close the running socket through the browser, a new connection works fine. Otherwise, this is what happens via cURL: 

```
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost.com (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
^C
```

On the other hand, if I change the `while loop` to `match` instead, like so:

```
match h2.accept().await {
         Some(request) => { ... },
         None => {}
}
```
And add the rest of the logic from the first snippet, nothing happens, I get an error in the browser. Via cURL this happens: 
```*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost.com (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=localhost.com
*  start date: Nov 21 13:19:47 2019 GMT
*  expire date: Feb 19 13:19:47 2020 GMT
*  subjectAltName: host "localhost.com" matched cert's "localhost.com"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fd63000b200)
GET /index.html HTTP/2
Host: localhost.com
User-Agent: curl/7.54.0
Accept: /


* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* Unexpected EOF
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, Client hello (1):
curl: (56) Unexpected EOF
```

And it's just ERR_CONNECTION_CLOSED in the browser. I have note yet used Wireshark to check what happens with the stream itself, but I though I check up with You so you could maybe provide me with some guidelines, I may be missing something?

This entire piece of code is inside `tokio::span(async move {...`, where I create the TcpListener, setup `tokio_rustls`, etc..

So I am not exactly sure why is there a difference between the implementation with the `while` loop and the `match` case, but I believe there is something, perhaps obvious, I am missing here. I will be grateful if You could provide me with some pointers on what to do next or how to fix these.

**Comment by seanmonstar** [comment]
Hm, does it work without TLS? The code you mentioned looks similar to what's in `examples/server.rs`...

**Comment by tuboythers** [comment]
I have not really tried using it insecurely, I aim to support H2 only over TLS, but I can try it over non-TLS socket later once I get a hold of it.

It's pretty similar to the `server.rs` example, yes, with the only difference that the `listener.accept()` and the h2 handshake `connection.accept()` are both in the same executor `tokio::spawn(async move { ...`, I have not yet separated them. Maybe it has something to do with that, should I separate them?

Based on what you I've provided do you see if I missed something?

**Comment by seanmonstar** [comment]
Oh, that's key information! Yes, if you have the listener and connection in the same task, then once the task is busy with a connection, it can't also be checking the listener.

Generally, you'll want to spawn a new task for a logical chunk of work.

**Comment by tuboythers** [comment]
I though that might be it, but I am too vaguely familiar with tokio, so I am still experimenting. Thank you @seanmonstar, I will try it a bit later and will see how that goes.

Please do not close this yet, I will close it automatically later on if everything works as expected.

**Comment by tuboythers** [comment]
You were right, it worked! 

Interestingly though, I still cannot use the `match` case, it still returns ERR_CONNECTION_RESET, but the loop `while let Some(request) = h2.accept().await {` works just as intended, strange!

Would you happen to know why, or is that expected?

Otherwise, you can close this one, thank you for you help! :)

**Comment by seanmonstar** [comment]
Without the loop, you're queuing frames be sent on the connection, but then dropping it before they can be written. You'd need another call to `accept` to have the frames written.
